cpp/src/io/json/host_tree_algorithms.cu

```diff
@@ -805,4 +802,595 @@ void scatter_offsets(
+                                                 d_unique_col_ids,
+                                                 d_max_row_offsets,
+                                                 column_names,
+                                                 row_array_parent_col_id,
+                                                 is_array_of_arrays,
+                                                 options,
+                                                 stream,
+                                                 mr);
+  if (ignore_vals.empty()) return;
+  scatter_offsets(tree,
+                  col_ids,
+                  row_offsets,
+                  node_ids,
+                  sorted_col_ids,
+                  d_column_tree,
+                  ignore_vals,
+                  columns,
+                  stream);
+}
+
+std::pair<thrust::host_vector<uint8_t>, hashmap_of_device_columns> build_tree(
+  device_json_column& root,
+  host_span<uint8_t const> is_str_column_all_nulls,
+  tree_meta_t& d_column_tree,
+  device_span<NodeIndexT const> d_unique_col_ids,
+  device_span<size_type const> d_max_row_offsets,
+  std::vector<std::string> const& column_names,
+  NodeIndexT row_array_parent_col_id,
+  bool is_array_of_arrays,
+  cudf::io::json_reader_options const& options,
+  rmm::cuda_stream_view stream,
+  rmm::device_async_resource_ref mr)
+{
+  bool const is_enabled_lines                 = options.is_enabled_lines();
+  bool const is_enabled_mixed_types_as_string = options.is_enabled_mixed_types_as_string();
+  auto unique_col_ids = cudf::detail::make_host_vector_async(d_unique_col_ids, stream);
+  auto column_categories =
+    cudf::detail::make_host_vector_async(d_column_tree.node_categories, stream);
+  auto const column_parent_ids =
+    cudf::detail::make_host_vector_async(d_column_tree.parent_node_ids, stream);
+  auto column_range_beg =
+    cudf::detail::make_host_vector_async(d_column_tree.node_range_begin, stream);
+  auto const max_row_offsets = cudf::detail::make_host_vector_async(d_max_row_offsets, stream);
+  auto num_columns           = d_unique_col_ids.size();
+
+  auto to_json_col_type = [](auto category) {
+    switch (category) {
+      case NC_STRUCT: return json_col_t::StructColumn;
+      case NC_LIST: return json_col_t::ListColumn;
+      case NC_STR: [[fallthrough]];
+      case NC_VAL: return json_col_t::StringColumn;
+      default: return json_col_t::Unknown;
+    }
+  };
+
+  auto initialize_json_columns = [&](auto i, auto& col_ref, auto column_category) {
+    auto& col = col_ref.get();
+    if (col.type != json_col_t::Unknown) { return; }
+    if (column_category == NC_ERR || column_category == NC_FN) {
+      return;
+    } else if (column_category == NC_VAL || column_category == NC_STR) {
+      col.string_offsets.resize(max_row_offsets[i] + 1, stream);
+      col.string_lengths.resize(max_row_offsets[i] + 1, stream);
+      thrust::uninitialized_fill(
```

Comment:
Can we merge the two thrust calls -
```suggestion
      thrust::uninitialized_fill(rmm::exec_policy_nosync(stream), 
      thrust::make_zip_iterator(col.string_offsets.begin(), col.string_lengths.begin()),
      thrust::make_zip_iterator(col.string_offsets.end(), col.string_lengths.end()), 
      0);
```

Reply:
Can't use uninitialized_fill here. this will replaced with  single `fill` call.
`error: a reference of type "thrust::tuple<int, int> &" (not const-qualified) cannot be initialized with a value of type "const thrust::detail::tuple_of_iterator_references<int &, int &>"`

 I benchmarked this to see if merging them is faster. It is.

## Benchmark Results

## uninitialized_fill_separate

### [0] NVIDIA RTX 5880 Ada Generation

|       size        | Samples |  CPU Time  | Noise  |  GPU Time  | Noise |  Elem/s  | GlobalMem BW | BWUtil |
|-------------------|---------|------------|--------|------------|-------|----------|--------------|--------|
|       2^10 = 1024 |  52080x |  12.852 us | 34.77% |   9.602 us | 6.60% | 106.649M | 853.189 MB/s |  0.09% |
|      2^14 = 16384 |  52640x |  13.144 us | 39.49% |   9.499 us | 6.71% |   1.725G |  13.798 GB/s |  1.44% |
|     2^18 = 262144 |  49504x |  13.361 us | 32.79% |  10.100 us | 6.42% |  25.954G | 207.632 GB/s | 21.63% |
|    2^22 = 4194304 |  12640x |  43.291 us |  9.49% |  39.557 us | 1.28% | 106.031G | 848.246 GB/s | 88.35% |
|   2^26 = 67108864 |    830x | 605.977 us |  0.59% | 602.436 us | 0.10% | 111.396G | 891.167 GB/s | 92.82% |
| 2^30 = 1073741824 |     52x |   9.638 ms |  0.04% |   9.634 ms | 0.01% | 111.453G | 891.628 GB/s | 92.87% |

## fill_separate

### [0] NVIDIA RTX 5880 Ada Generation

|       size        | Samples |  CPU Time  | Noise  |  GPU Time  | Noise |  Elem/s  | GlobalMem BW | BWUtil |
|-------------------|---------|------------|--------|------------|-------|----------|--------------|--------|
|       2^10 = 1024 |  51376x |  12.982 us | 34.17% |   9.733 us | 6.23% | 105.211M | 841.689 MB/s |  0.09% |
|      2^14 = 16384 |  52672x |  13.137 us | 38.84% |   9.494 us | 6.79% |   1.726G |  13.806 GB/s |  1.44% |
|     2^18 = 262144 |  49504x |  13.359 us | 32.72% |  10.103 us | 6.45% |  25.948G | 207.585 GB/s | 21.62% |
|    2^22 = 4194304 |  12640x |  43.340 us |  9.96% |  39.571 us | 1.57% | 105.993G | 847.946 GB/s | 88.32% |
|   2^26 = 67108864 |    830x | 606.034 us |  0.60% | 602.505 us | 0.12% | 111.383G | 891.064 GB/s | 92.81% |
| 2^30 = 1073741824 |     52x |   9.638 ms |  0.04% |   9.634 ms | 0.01% | 111.453G | 891.625 GB/s | 92.87% |

## fill_merged

### [0] NVIDIA RTX 5880 Ada Generation

|       size        | Samples |  CPU Time  | Noise  |  GPU Time  | Noise  |  Elem/s  | GlobalMem BW | BWUtil |
|-------------------|---------|------------|--------|------------|--------|----------|--------------|--------|
|       2^10 = 1024 |  72112x |  10.185 us | 47.99% |   6.934 us |  9.39% | 147.685M |   1.181 GB/s |  0.12% |
|      2^14 = 16384 |  74864x |  10.311 us | 55.96% |   6.679 us | 10.52% |   2.453G |  19.624 GB/s |  2.04% |
|     2^18 = 262144 |  66352x |  10.792 us | 44.19% |   7.536 us |  7.96% |  34.786G | 278.289 GB/s | 28.99% |
|    2^22 = 4194304 |  12528x |  43.754 us |  9.94% |  39.957 us |  1.55% | 104.971G | 839.764 GB/s | 87.47% |
|   2^26 = 67108864 |    828x | 607.637 us |  0.59% | 604.102 us |  0.10% | 111.089G | 888.708 GB/s | 92.56% |
| 2^30 = 1073741824 |     52x |   9.640 ms |  0.04% |   9.636 ms |  0.01% | 111.427G | 891.419 GB/s | 92.85% |

</details>